File: edukate-common/src/main/java/io/github/sanyavertolet/edukate/common/utils/HttpHeadersUtils.java
package io.github.sanyavertolet.edukate.common.utils;


import io.github.sanyavertolet.edukate.common.users.EdukateUserDetails;
import io.github.sanyavertolet.edukate.common.users.UserRole;
import io.github.sanyavertolet.edukate.common.users.UserStatus;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;

import java.util.List;
import java.util.Objects;

@Slf4j
@UtilityClass
public class HttpHeadersUtils {
    public static String getSingleHeader(HttpHeaders headers, String headerName) {
        List<String> headerCandidates = headers.get(headerName);
        if (headerCandidates != null && headerCandidates.size() == 1) {
            return headerCandidates.getFirst();
        }
        log.trace("Header {} is not provided: skipping pre-authenticated edukate-user authentication", headerName);
        return null;
    }

    public static void populateHeaders(HttpHeaders httpHeaders, EdukateUserDetails edukateUserDetails) {
        Objects.requireNonNull(httpHeaders, "HttpHeaders must not be null");
        httpHeaders.set(AuthHeaders.AUTHORIZATION_ID.headerName, edukateUserDetails.getId());
        httpHeaders.set(AuthHeaders.AUTHORIZATION_NAME.headerName, edukateUserDetails.getUsername());
        httpHeaders.set(AuthHeaders.AUTHORIZATION_STATUS.headerName, edukateUserDetails.getStatus().toString());
        String rolesString = UserRole.toString(edukateUserDetails.getRoles());
        httpHeaders.set(AuthHeaders.AUTHORIZATION_ROLES.headerName, rolesString);
    }

    public static EdukateUserDetails toEdukateUserDetails(HttpHeaders headers) {
        String id = getSingleHeader(headers, AuthHeaders.AUTHORIZATION_ID.headerName);
        String name = getSingleHeader(headers, AuthHeaders.AUTHORIZATION_NAME.headerName);
        String roles = getSingleHeader(headers, AuthHeaders.AUTHORIZATION_ROLES.headerName);
        String status = getSingleHeader(headers, AuthHeaders.AUTHORIZATION_STATUS.headerName);

        if (id == null || name == null || roles == null || status == null) {
            return null;
        }

        return new EdukateUserDetails(id, name, UserRole.fromString(roles), UserStatus.valueOf(status), null);
    }
}

File: edukate-common/src/main/kotlin/io/github/sanyavertolet/edukate/common/utils/HttpHeadersUtils.kt
package io.github.sanyavertolet.edukate.common.utils

import io.github.sanyavertolet.edukate.common.users.EdukateUserDetails
import io.github.sanyavertolet.edukate.common.users.UserRole
import io.github.sanyavertolet.edukate.common.users.UserStatus
import org.slf4j.LoggerFactory
import org.springframework.http.HttpHeaders

fun populateHeaders(httpHeaders: HttpHeaders, edukateUserDetails: EdukateUserDetails) {
    httpHeaders.apply {
        set(AuthHeaders.AUTHORIZATION_ID.headerName, edukateUserDetails.id.toString())
        set(AuthHeaders.AUTHORIZATION_NAME.headerName, edukateUserDetails.username)
        set(AuthHeaders.AUTHORIZATION_STATUS.headerName, edukateUserDetails.status.toString())
        set(AuthHeaders.AUTHORIZATION_ROLES.headerName, UserRole.listToString(edukateUserDetails.roles))
    }
}

fun HttpHeaders.toEdukateUserDetails(): EdukateUserDetails? {
    val id = get(AuthHeaders.AUTHORIZATION_ID.headerName)?.lastOrNull()
    val name = get(AuthHeaders.AUTHORIZATION_NAME.headerName)?.lastOrNull()
    val rolesString = get(AuthHeaders.AUTHORIZATION_ROLES.headerName)?.lastOrNull()
    val statusString = get(AuthHeaders.AUTHORIZATION_STATUS.headerName)?.lastOrNull()

    @Suppress("ComplexCondition")
    if (id == null || name == null || rolesString == null || statusString == null) {
        logger.trace("Authentication headers are is not provided: skipping pre-authenticated edukate user")
        return null
    }

    val numericId = id.toLongOrNull() ?: return null
    return EdukateUserDetails(numericId, name, UserRole.fromString(rolesString), UserStatus.valueOf(statusString), "")
}

private val logger = LoggerFactory.getLogger("HttpHeadersUtils")
